go/vt/vtgate/engine/specialized.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2025 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package engine
+
+import (
+	"bytes"
+	"context"
+	"strings"
+
+	"vitess.io/vitess/go/slice"
+
+	"vitess.io/vitess/go/sqltypes"
+	querypb "vitess.io/vitess/go/vt/proto/query"
+)
+
+// Condition lists arguments that need to point to the same value for the plan to be used
+type Condition struct {
+	// these arguments need to have equal values for this condition to be true
+	A, B string
+}
+
+// PlanSwitcher chooses between a baseline and an optimized plan based on conditions
+type PlanSwitcher struct {
+	noTxNeeded
+
+	// Conditions lists the conditions that need to be true for the optimized plan to be used
+	Conditions  []Condition
+	Baseline    Primitive
+	BaselineErr error
+	Optimized   Primitive
+}
+
+func (s *PlanSwitcher) RouteType() string {
+	return "PlanSwitcher"
+}
+
+func (s *PlanSwitcher) GetKeyspaceName() string {
+	if s.Baseline != nil {
+		return s.Baseline.GetKeyspaceName()
+	}
+	return s.Optimized.GetKeyspaceName()
```

Comment:
Shouldn't it be the opposite? e.g.:
```go
	if s.Optimized != nil {
		return s.Optimized.GetKeyspaceName()
	}
	return s.Baseline.GetKeyspaceName()
```

Same question for `GetTableName()`.

Reply:
Mhm I realized that we create `PlanSwitcher` only once, and `Optimized` is always set at that time.

Reply:
We might not always run the optimized version so better to give the details from baseline plan